### PR TITLE
JAMES-2661 WebAdmin API for EventDeadLetters

### PR DESCRIPTION
**Pull request by chibenwa**
The trickiest part is about group parsing...

Then I implemented the routes for:
 - Listing groups
 - Retrieving events that a given group failed to execute
 - Retrieving an event JSON - I reused the Scala code here...
 - Deleting an event

I do think we miss some documentation on event json structure...

### CODE REVIEW

File: mailbox/api/src/test/java/org/apache/james/mailbox/events/GroupTest.java [hunk 1]
```diff
@@ -75,4 +76,41 @@ void asStringShouldReturnFqdnByDefault() {
...
+
+    @Test
+    void deserializeShouldReturnGroupWhenGenericGroup() throws Exception {
+        assertThat(Group.deserialize("org.apache.james.mailbox.events.GenericGroup-abc"))
+            .isEqualTo(new GenericGroup("abc"));
+    }
+
+    @Test
+    void deserializeShouldReturnGroupWhenEmptyGenericGroup() throws Exception {
+        assertThat(Group.deserialize("org.apache.james.mailbox.events.GenericGroup-"))
```
[the remark this answers is not in the document]
Comment (trantienduchn) [reply mailbox/api/src/test/java/org/apache/james/mailbox/events/GroupTest.java]: I think we need to add some tests for passing null or empty string to the `deserialize()` method

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/Group.java [hunk 2]
```diff
@@ -21,7 +21,27 @@
...
+    public static Group deserialize(String serializedGroup) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
+        if (serializedGroup.startsWith(GenericGroup.class.getName() + GenericGroup.DELIMITER)) {
+            return new GenericGroup(serializedGroup.substring(GenericGroup.class.getName().length() + 1));
+        }
+        return loadGroup(serializedGroup);
+    }
+
+    private static Group loadGroup(String serializedGroup) throws ClassNotFoundException, InstantiationException, IllegalAccessException {
+        ClassLoader classLoader = Group.class.getClassLoader();
+        Class<?> aClass = classLoader.loadClass(serializedGroup);
```
Comment (trantienduchn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/Group.java]: Can be replaced by `Class.forName()`. It saves one line :)

File: src/site/markdown/server/manage-webadmin.md [hunk 3]
```diff
@@ -2413,6 +2414,88 @@ The scheduled task will have the following type `MessageIdReIndexingTask` and th
...
+ - [Getting event details](#Getting_event_details)
+ - [Deleting an event](#Deleting_an_event)
+ - [Rescheduling group execution](#Rescheduling_group_execution)
+
+### Listing groups
+
+This endpoint allows discovering the list of groups.
+
+```
+curl -XGET http://ip:port//events/deadLetter/groups
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: `ip:port//` has two slashes, similar to other examples

File: mailbox/api/src/test/java/org/apache/james/mailbox/events/GroupTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/events/GroupTest.java]: > what happens when I pass "org.apache.james.mailbox.events.GenericGroup" without "-"?

There is an exception - and a test about that

> I think we need to add some tests for passing null or empty string to the deserialize() method

Will do!

File: mailbox/api/src/main/java/org/apache/james/mailbox/events/Group.java [hunk 4]
```diff
@@ -21,7 +21,27 @@
 
 import java.util.Objects;
 
+import com.google.common.base.Preconditions;
+
 public class Group {
+    public static Group deserialize(String serializedGroup) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
+        if (serializedGroup.startsWith(GenericGroup.class.getName() + GenericGroup.DELIMITER)) {
```
Comment (hoangdat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/events/Group.java]: Can you explain to me: 
- What `GenericGroup` implies to?
- Why the `asString` format of `GenericGroup` is different with other?

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/events/Group.java]: Sure...

Basically, all groups have a different class. `GroupMailboxListener` typically exposes a `getDefaultGroup` method that allows retrieving the group to be used for this listener. Here it's easy: each listener exposes a group super class with no field. ClassName is used for equality.

However, we end up registrating several time the same listener class, and expect each listener to get a distinct group. This is for instance the case for `Quota Mailing`. We may want to set up a first listener to `trigger a 80% information` and another `trigger a 98% warning`... The above approach don't work: we need 2 different groups despite it being the same class. So we decided to configure it (James 1 & James 2 needs to use the same group for 80%, for 98%, etc...).  The `GenericGroup` holds the configured string allowing to get distinct execution groups.

So, if we get a Generic group, the format is `FQDN-configured_string` while in the generic case it is just `FQDN`

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        void getGroupsShouldReturnGroupsOfContainedEventsWithoutDuplicates() {
+            deadLetters.store(new EventBusTestFixture.GroupA(), EVENT_1).block();
+            deadLetters.store(new EventBusTestFixture.GroupA(), EVENT_2).block();
+
+            when()
+                .get("/events/deadLetter/groups")
+            .then()
+                .statusCode(HttpStatus.OK_200)
+                .contentType(ContentType.JSON)
+                .body(".", contains(EventBusTestFixture.GroupA.class.getName()));
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java]: Don't we have a kind of `containsOnly`?

File: src/site/markdown/server/manage-webadmin.md [hunk 6]
```diff
@@ -2413,6 +2414,88 @@ The scheduled task will have the following type `MessageIdReIndexingTask` and th
 
 Warning: During the re-indexing, the result of search operations might be altered.
 
+## Event Dead letter
+
+The EventBus allows to register 'group listeners' that are called in a (potentially) distributed fashion. These group
+listeners enable the implementation of some advanced mailbox manager feature like indexing, spam reporting, quota management
+and the like.
+
+Upon exceptions, a bounded number of retries is performed (with exponential backoff delays). If after those retries the listener is still
```
Comment (aduprat) [inline src/site/markdown/server/manage-webadmin.md]: s/ is performed / are performed

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java [hunk 7]
```diff
@@ -207,4 +225,54 @@ void listEventsShouldReturnAllEvents() {
...
     }
+
+    @Nested
+    class GetEvent {
+        @Test
+        void getEventShouldReturnEvent() {
+            deadLetters.store(new EventBusTestFixture.GroupA(), EVENT_1).block();
+
+            String response = when()
+                .get("/events/deadLetter/groups/" + new EventBusTestFixture.GroupA().asString() + "/events/" + UUID_1)
```
Comment (hoangdat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/EventDeadLettersRoutesTest.java]: `new EventBusTestFixture.GroupA().asString()` should we make it constant?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            required = true,
+            name = "group",
+            paramType = "path parameter",
+            dataType = "String",
+            defaultValue = "none",
+            value = "Compulsory. Needs to be a valid group name")
+    })
+    @ApiResponses(value = {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: It is a 400 if the group can not be deserialized

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        deadLetters.remove(group, eventId).block();
+        response.status(HttpStatus.NO_CONTENT_204);
+        return response;
+    }
+
+    private Group parseGroup(Request request) {
+        String groupAsString = request.params(GROUP_PARAM);
+        try {
+            return Group.deserialize(groupAsString);
+        } catch (Exception e) {
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: With this, you never get a 404 being thrown back as a response if Group doesn't exist. Is this acceptable? (Being defined in the task 404 when group not found)

Comment (Arsnael) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: Been fixed while I was reviewing, ignore this !

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: @aduprat : 404 has been defined and added as a default route exception in webadmin server since alias story, might be why you think it's missing !

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        deadLetters.remove(group, eventId).block();
+        response.status(HttpStatus.NO_CONTENT_204);
+        return response;
+    }
+
+    private Group parseGroup(Request request) {
+        String groupAsString = request.params(GROUP_PARAM);
+        try {
+            return Group.deserialize(groupAsString);
+        } catch (Group.GroupDeserializationException e) {
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: After reflexion, still the same. Do we take as granted that a non existing group (which raises the exception on the deserialization) returns 400? Should modify the description in the ticket then, as we have a 404 defined in it for group not registered

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java [hunk 11]
```diff
@@ -143,6 +143,7 @@ public void define(Service service) {
     @ApiResponses(value = {
         @ApiResponse(code = HttpStatus.OK_200, message = "OK - returns an event detail", response = Event.class),
         @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Invalid group name or event id"),
+        @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "No message with this eventId"),
```
Comment (hoangdat) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: should `No event ...` instead of `No message`?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java (hunk 10 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/EventDeadLettersRoutes.java]: Well I do think we don't need to stick on the description of the ticket for the following reason:

 - A "not found group" means the FQDN specified is not valid. 
 - A group is valid will automatically "be found" and will never return a 404
 - So, the root cause of a possible 404 is not that a resopurce do not exist, but that the resources asked for **is not valid.

I really don't think a 404 describes this better. A 400 do.

**Comment by chibenwa** [comment]
Merged
